cmd_export.go

```diff
@@ -52,6 +52,16 @@ func exportCommand(ctx context.Context, env Env, args []string, config *Config)
 		target = args[1]
 	}
 
+	// PATH is the "logical" path manipulated by stdlib and .envrc
+	// DIRENV_PLATFORM_PATH is the OS-specific path needed for invoking programs.
+	// Anywhere but Windows, these are both the same.
+	env[DIRENV_PLATFORM_PATH] = env["PATH"]
```

Comment:
The current implementation seems quite tricky to get right. The developer has to know all the execution paths and make sure that the env is added or removed at the right places.

Maybe it would work to add it to env.CleanContext() and in the IgnoredKeys?

Reply:
No.  I thought I could do that, but it means that `direnv exec` wouldn't work.  The platform path has to be diffed, so it can't be ignored.  (Otherwise, `direnv exec doesn't` know what the platform path should be when the current .envrc is unloaded.)

So, what you see now is currently the simplest logic that would still let direnv exec function, while also being (relatively) straightforward to reason about.

Specifically, the invariant is that `DIRENV_PLATFORM_PATH` is never directly visible as a variable *outside* of the direnv process, but must always be present *inside* of it. So `ToGoEnv` replaces `PATH` with it and removes it.  And in a handful of other places where it would be visible to the parent or child process, it's deleted.  But, it must *not* be removed at any other time.  As soon as an env-affecting command starts, it sets it from PATH, and there it stays until there's a chance of it being exposed to another program.

TBH, the hairiness of where those deletes are has more to do with existing organizational issues in the code, than to do with the changes themselves.  There is a lot of semi-invisible state coupling taking place in export, not counting the indirect state coupling *across* different commands!

I almost think it'd be easier to replace direnv dump with some bash code; an awful lot of the hairyness for Windows has to do with it in one way or another.  Dumping anything but the gzenv format would be pretty easy to do from bash in the stdlib, and it would fix the various piping, tempfile paths, etc. issues I'm having with the test suite.  But of course there are other problems with that approach...

Reply:
> TBH, the hairiness of where those deletes are has more to do with existing organizational issues in the code, than to do with the changes themselves. There is a lot of semi-invisible state coupling taking place in export, not counting the indirect state coupling across different commands!

Ouch. This rings very true :) Even as the main author I need to re-load all of this complexity every time I touch the main engine.

Reply:
Yeah.  I'm finding this is harder than it looks; notably, I haven't yet encountered this issue at runtime use of this version, but I'm pretty sure that watches on anything other than the original .envrc file aren't going to work, because the stdlib will pass a unix path to direnv watch, which being a Windows program, it won't understand.

(Ironically, going the other way is not necessarily a problem; Cygwin and MSYS programs can use Windows paths just fine, as long as they don't try to *manipulate* them in any way, which is why the initial call to `source_env` works fine despite direnv supplying a Windows path.)

Likewise, the stdlib's call to `direnv dotenv` would also need to convert the path it sends in.

I'm starting to think it makes sense to play around more with doing .direnvrc additions to patch the stdlib for cygwin/msys.

I think what I should do at this point is first do a PR *just* to fix the incorrect path in `direnv hook`, which I think we can agree is an uncontroversial change, since it makes it possible to *try* running direnv on Windows.  :wink:

Next, a PR for some changes to the stdlib to detect a cygwin or msys environment, so as to only pass Windows paths to direnv {dotenv,watch,dump,...}.

At that point, we can revisit how to address PATH, empty environment variables, and anything else that comes up.  The only real difficulty I see is that the path conversion changes to the stdlib can't really be tested by the test suite yet, except insofar as testing that they don't break non-Windows platforms.  ;-)

What do you think of that?  Or I suppose I could just do one PR with both the direnv hook fix and the stdlib changes for Windows path conversion?

(I'm thinking the path conversion change for the stdlib could just be something like defining a direnv function in the stdlib that runs "$direnv" "$@" on non-Windows, but on Windows rewrites any paths in the arguments and in `DIRENV_DUMP_FILE_PATH`.  Then replace all the current `$direnv` invocations with calls to that function instead.)

Alternately, I guess I could go ahead and do that in *this* PR, and see if that gets the test suite running, apart from empty vars.

Your thoughts?

Reply:
Another thought: `expand_path` can be portably implemented in pure bash, which would fix one subcommand right up, and allow dropping the `direnv expand_path` command.

The `realpath.absolute` function from [realpaths](https://github.com/bashup/realpaths) implements an all-bash substitute for the go `expandPath` function, just with arguments ordered the other way around (i.e. "[relto] path" rather then "path [relto]").  That would let us just drop `cmd_expand_path.go` entirely, incidentally making things faster on Windows since forks are expensive there.  (And the current way of using expand_path requires two forks and an exec, vs. just one fork and no execs with a bash version.)

In other words, by adding the `realpath.absolute` and `realpath.dirname` functions from realpaths (about 12 lines of code) to the stdlib, the new version of expand_path would look like this:

```shell
expand_path() {
  local REPLY; realpath.absolute ${2+"$2"} ${1+"$1"}; echo "$REPLY"
}
```

And then that's one less thing to figure out Windows path translations for.  ;-)

So in principle, this could be split to another prerequisite PR, i.e. replacing cmd_expand_path.go with the above version of expand_path.

Reply:
I think it's a good idea to take baby steps. Make the minimal changes in this PR so that things mostly work on Windows. Disable most of the tests for Windows and test in CI. Then expand.

The worst outcome, in my opinion, would be to aim for something perfect and run out of energy before it's finished. As long as it doesn't break backward-compatibility then it's fine.

+1 on replacing expand_path with a pure-bash implementation as long as it works in Bash 3.0 (thanks macOS...)

Reply:
PR #631 is bash 3.2 compatible.  The original raison d'etre for bashup/realpaths, as it happens, was to provide a realpath alternative for OS X without homebrew.  So it had to not only be pure bash, but pure bash 3.2.  I'm actually pretty hyper-aware of the differences between certain versions of bash, such as the horrible interaction of `set -u` and array variables that happens in bash 4.3, but not 3.2-4.2 or 4.4+.  :wink: